
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://thwacker.github.io/DiPPER2/">
      
      
      
        <link rel="next" href="reference/">
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.46">
    
    
      
        <title>DiPPER2 documentation</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.6f8fc17f.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#dipper2" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="DiPPER2 documentation" class="md-header__button md-logo" aria-label="DiPPER2 documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            DiPPER2 documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Documentation
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="DiPPER2 documentation" class="md-nav__button md-logo" aria-label="DiPPER2 documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    DiPPER2 documentation
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Documentation
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="." class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Documentation
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#synopsis" class="md-nav__link">
    <span class="md-ellipsis">
      Synopsis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    <span class="md-ellipsis">
      Features
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Requirements
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      Installation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    <span class="md-ellipsis">
      Usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dipper2-results-walkthrough" class="md-nav__link">
    <span class="md-ellipsis">
      DiPPER2 results walkthrough
    </span>
  </a>
  
    <nav class="md-nav" aria-label="DiPPER2 results walkthrough">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-results-folder-contains" class="md-nav__link">
    <span class="md-ellipsis">
      The results folder contains:
    </span>
  </a>
  
    <nav class="md-nav" aria-label="The results folder contains:">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#within-the-subfolders-the-following-files-can-be-found" class="md-nav__link">
    <span class="md-ellipsis">
      Within the subfolders, the following files can be found:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-information" class="md-nav__link">
    <span class="md-ellipsis">
      Other information:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contact" class="md-nav__link">
    <span class="md-ellipsis">
      Contact
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#acknowledgements" class="md-nav__link">
    <span class="md-ellipsis">
      Acknowledgements
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Reference
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#synopsis" class="md-nav__link">
    <span class="md-ellipsis">
      Synopsis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    <span class="md-ellipsis">
      Features
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Requirements
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      Installation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    <span class="md-ellipsis">
      Usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dipper2-results-walkthrough" class="md-nav__link">
    <span class="md-ellipsis">
      DiPPER2 results walkthrough
    </span>
  </a>
  
    <nav class="md-nav" aria-label="DiPPER2 results walkthrough">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-results-folder-contains" class="md-nav__link">
    <span class="md-ellipsis">
      The results folder contains:
    </span>
  </a>
  
    <nav class="md-nav" aria-label="The results folder contains:">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#within-the-subfolders-the-following-files-can-be-found" class="md-nav__link">
    <span class="md-ellipsis">
      Within the subfolders, the following files can be found:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-information" class="md-nav__link">
    <span class="md-ellipsis">
      Other information:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contact" class="md-nav__link">
    <span class="md-ellipsis">
      Contact
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#acknowledgements" class="md-nav__link">
    <span class="md-ellipsis">
      Acknowledgements
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<p><a href="https://github.com/ThWacker/DiPPER2/actions/workflows/codeql.yml"><img alt="CodeQL Advanced" src="https://github.com/ThWacker/DiPPER2/actions/workflows/codeql.yml/badge.svg" /></a> <a href="https://github.com/ThWacker/DiPPER2/actions/workflows/pylint.yml"><img alt="Python Lint and Code Quality" src="https://github.com/ThWacker/DiPPER2/actions/workflows/pylint.yml/badge.svg" /></a>
<a href="https://www.repostatus.org/#active"><img alt="Project Status: Active – The project has reached a stable, usable state and is being actively developed." src="https://www.repostatus.org/badges/latest/active.svg" /></a>
<img alt="Static Badge" src="https://img.shields.io/badge/unittest_code_coverage-78%25-yellow" />
<img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/ThWacker/DiPPER2" />
<a href="https://www.gnu.org/licenses/gpl-3.0"><img alt="License: GPL v3" src="https://img.shields.io/badge/License-GPLv3-blue.svg" /></a>
<img alt="" src="./logo/Dipper2_white.png" /> </p>
<h1 id="dipper2">DiPPER2</h1>
<p>DiPPER2:  Diagnostic Primer Picking and Evaluation pipeline for Reliability and Reproducibility</p>
<blockquote>
<p><strong>+++++++++++++++++++++++++++++NOTE - IMPORTANT:+++++++++++++++++++++++++++++++</strong></p>
<p><strong>This is a work in progress which has not been extensively unit tested and will likely undergo substantial refactoring in the near future.</strong></p>
<p><strong>+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++</strong></p>
</blockquote>
<h2 id="synopsis">Synopsis</h2>
<p><strong><em>This pipeline and modules are meant to facilitate reliable and reproducible finding of diagnostic targets and to make picking primers for those targets as user-friendly as possible. The approach taken is a phylogeny-driven and clade-specific approach.</em></strong> </p>
<p>The pipeline, once functional, automatically finds unique genomic regions for a target species, pathovar or race with moderate user intervention, automatically picks primers for these regions that are amenable to both conventional PCR and quantitative PCR, and finally tests them for specificity and sensitivity.</p>
<h2 id="overview">Overview</h2>
<p>DiPPER2 is a tool to find unique genomic regions of target species, build primers (conv. or qPCR) for them, validate them <em>in silico</em>, define the identity of the unique genomic regions (either by homology or genome coordinates) and produce a user-friendly html report (and a more machine-readable .txt report).</p>
<h2 id="features">Features</h2>
<p>Key features of DiPPER2:</p>
<ul>
<li>generates a html and .txt report with the primers, the identity of the region the primers target and information of specificity and sensitivity.</li>
<li>generates .txt files (in fasta format) with primers for the unique genomic regions found</li>
<li>generates fasta files with the sequence of the unique genomic regions the primers target</li>
<li>generates .txt files that contain information on the primer's characteristics (Tm, GC, product size <em>etc</em>.)</li>
<li>generates .bed files of unique genomic regions that cannot be identified by homology (blastx). These can be used in IGV to find where the unique genomic region falls in a reference genome (either provided, or longest)</li>
<li>generates .bed files for the <em>in silico</em> tests of primers</li>
</ul>
<p>For more details on the outputs and the folder structure, compare the 'DiPPER2 results walkthrough' section.</p>
<h2 id="requirements">Requirements</h2>
<p>The following programs need to be in path:</p>
<ul>
<li><a href="https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html">BLAST+</a></li>
<li><a href="https://bioinf.shenwei.me/seqkit/download/">seqkit</a></li>
<li><a href="https://github.com/EvolBioInf/fur">FUR</a></li>
<li><a href="https://primer3.org/releases.html">primer3_core</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/home/tools/">E-utilities</a></li>
</ul>
<p>DiPPER2 was developed using python 3.9 and 3.12. It runs stably on both version 3.9 and 3.12, but a python version &gt;=3.12 is recommended.</p>
<p>DiPPER2 automatically generates a python environment and installs the necessary python modules defined in the requirements file.</p>
<p>Seqkit, BLAST+ and primer3 can be installed via conda. primer3 can also be installed using homebrew. </p>
<h2 id="installation">Installation</h2>
<p>Assuming all the required programs listed in Requirements are installed and in path, DiPPER2 can be installed by cloning the repo:</p>
<pre><code class="language-Bash">git clone https://github.com/ThWacker/DiPPER2.git
</code></pre>
<h2 id="usage">Usage</h2>
<h2 id="dipper2-results-walkthrough">DiPPER2 results walkthrough</h2>
<h3 id="the-results-folder-contains">The results folder contains:</h3>
<ul>
<li>
<p><strong>subfolders:</strong></p>
<ul>
<li>FUR.P3.PRIMERS</li>
<li>FUR.P3.PRIMERS/primer_data</li>
<li>FUR.P3.PRIMERS/in_silico_tests</li>
<li>FUR.P3.TARGETS</li>
</ul>
</li>
<li>
<p><strong>files:</strong></p>
<ul>
<li>Results.txt and Results.html: </br> the results summary files</li>
<li><code>&lt;Prefix&gt;</code>.targets.txt and <code>&lt;Prefix&gt;</code>.neighbours.txt: files listing the accessions/ files used as targets and neighbours</li>
<li><code>&lt;Prefix&gt;</code>.FUR.db.out.txt: </br>  the FUR results file (a multifasta file)</li>
<li><Prefix>.FUR.db.out.primers.txt: </br>  the FUR results file reformatted in the format required by Primer3 as an input</li>
<li><code>&lt;Prefix&gt;</code>.FUR.db.out.primers.primer3_out.txt: </br>  the primer3 output results file</li>
<li>FUR_Summary_output.txt: </br> FUR output to STDERR which summarizes how many sequences were retained after each processing step, including the length in bp and the number of Ns</li>
<li><em>optional:</em></br>  bed files generated by seqkit locate</li>
</ul>
</li>
</ul>
<h4 id="within-the-subfolders-the-following-files-can-be-found">Within the subfolders, the following files can be found:</h4>
<ul>
<li><strong>within FUR.P3.PRIMERS:</strong> </br>
    Primer fasta files (.txt ending, but fasta formatted). Primers are numbered. The number is the unique identity of the primer.</li>
<li><strong>within FUR.P3.PRIMERS/primer_data:</strong></br>
    Here, textfiles with information about Tm, amplicon length, GC etc are found. These follow the Primer3 conventions, please compare here for explanation: (<a href="https://primer3.org/manual.html#outputTags">https://primer3.org/manual.html#outputTags</a>)</li>
<li><strong>within FUR.P3.PRIMERS/in_silico_tests</strong> </br>
    Here, the results of the in silico PCR tests, performed using seqkit locate, can be found. Files names with 'target' in the name are files used for sensitivity testing, running an in silico PCR against the concatenated targets.
    File names with 'neighbour' in the name are files used for specificity testing, running an in silico PCR against the concatenated neighbours. The 'm' in the name refers to the number of allowed mismatches in the primer</li>
<li><strong>within FUR.P3.TARGETS:</strong> </br>
    Here, we find target files with 'Target' in the name, followed by a number that matches the unique identifier of the primer pair, which are fasta files containing the target sequence.
    Also, blastx results are found here. These are tab-separated and the headers of the blastx files are "qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore"</li>
</ul>
<h3 id="other-information">Other information:</h3>
<p>All relevant information are found in the results files, which will give you an idea whether a primer passed or failed. Currently, the results files do not contain amplicon lengths or Tms etc, please refer to the txt files in the FUR.P3.PRIMERS/primer_data folder for that. </p>
<p>Up to 4 primers are picked, but sometimes less than 4 primers are generated by Primer3. Having less than 4 primers in the results does not mean the run/ pipeline did not complete.</p>
<p>Currently, the default target parameters for conventional primers are:</br> 
<code>primMinTm=58 primOptTm=60 primMaxTm=62 inMinTm=63 inOptTm=65 inMaxTm=67 prodMinSize=200 prodMaxSize=1000 Oligo=0</code></p>
<p>The default target parameters for qPCR primers are:</br> 
<code>primMinTm=58 primOptTm=60 primMaxTm=62 inMinTm=63 inOptTm=65 inMaxTm=67 prodMinSize=100 prodMaxSize=200</code></p>
<p>This means that conventional primers have a target optimum Tm of 60 and a product size of 200-1000 and qPCR primers a Tm of 60 and an amplicon size of 100-200. For qPCR primers an internal probe is also picked with an optimal Tm of 65.</p>
<h2 id="contact">Contact</h2>
<p>Either open an issue in this repo or contact Theresa Wacker at t[dot]wacker2[at]exeter[dot]ac[dot]uk.</p>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>DiPPER2 uses FUR, primer3, seqkit, BLAST+ and eutilies, all of which have been published. When using DiPPER2, please also cite:</p>
<p>Camacho, C., Coulouris, G., Avagyan, V., Ma, N., Papadopoulos, J., Bealer, K. and Madden, T.L. (2009). BLAST+: architecture and applications. BMC Bioinformatics, 10(1), p.421.</p>
<p>Haubold, B., Klötzl, F., Hellberg, L., Thompson, D. and Cavalar, M. (2021). Fur: Find unique genomic regions for diagnostic PCR. Bioinformatics, 37(15), pp.2081–2087. doi:https://doi.org/10.1093/bioinformatics/btab059.</p>
<p>Kans, J. (2016). Entrez Direct: E-utilities on the UNIX Command Line. [online] Available at: https://www.ncbi.nlm.nih.gov/books/NBK179288/ [Accessed 4 Dec. 2024].</p>
<p>Shen, W., Le, S., Li, Y. and Hu, F. (2016). SeqKit: A Cross-Platform and Ultrafast Toolkit for FASTA/Q File Manipulation. PLOS ONE, 11(10), p.e0163962. doi:https://doi.org/10.1371/journal.pone.0163962.</p>
<p>Untergasser, A., Cutcutache, I., Koressaar, T., Ye, J., Faircloth, B.C., Remm, M. and Rozen, S.G. (2012). Primer3—new capabilities and interfaces. Nucleic Acids Research, 40(15), pp.e115–e115. doi:https://doi.org/10.1093/nar/gks596.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": ".", "features": [], "search": "assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.83f73b43.min.js"></script>
      
    
  </body>
</html>